[TF] TensorGroup and TensorArrayProtocol conformance derivation

eaplatanios opened the pull request:
Resolves TF-259.

~~The derivation for `TensorGroup` is only missing the initializer currently, but this should be easy to add after I'm done with the `TensorArrayProtocol` derivation. My main issue currently is that even though I have added the `_unpackTensorHandles` derivation, the compiler does not seem to use it and keeps throwing conformance errors when that one is missing. UPDATE: I figured this out.~~

~~Once that is resolved I will add the `TensorGroup` initializer as it's very similar to the derivation of `_unpackTensorHandles`.~~

~~EDIT: Also, why was there a default implementation for `_unknownShapeList`? I removed it because this prevents the type checker from deriving it.~~

eaplatanios committed:
Added support for sampling from a Beta distribution.

eaplatanios committed:
Formatting edits.

eaplatanios committed:
Bug fix.

rxwei committed:
Update Random.swift

eaplatanios committed:
Merged upstream changes.

bgogul committed:
Use tracing to implement some of the dataset APIs (#22766)

Commit:
update lldb commit hash

rxwei committed:
[AutoDiff] Handle multiple `@differentiable` attributes in TBDGen. (#22809)

marcrasi committed:
tf-239 (#22767)

bgogul committed:
Allow Dataset.map to return an element of different type. (#22820)

Commit:
Fixed a bug in python object refcounts, where isType() consumed refcount for (#22816)

type objects like `PyBool_Type`, triggering segfaults after repeated invocation.

dan-zheng committed:
[AutoDiff] TF-199: Set `ParamDecl` types in `createEntryArguments`. (#22829)

The `createEntryArguments` helper creates arguments for an entry basic block
based on the type of its enclosing function.

Dummy `ParamDecl`s created in `createEntryArguments` must explicitly have
their type set to prevent crashes during optimizations.

Resolves TF-199.

eaplatanios committed:
Started working on derived conformances for TensorGroup.

dan-zheng committed:
[AutoDiff] TF-246: Fix adjoint buffer calculation for projections. (#22807)

* [AutoDiff] TF-246: Fix adjoint buffer calculation for projections.

Previously, adjoint buffers for projection instructions were newly
allocated buffers in the adjoint - this led to some derivative
miscalculations.

Now, adjoint buffers for address-to-address projections are projections
in the underlying buffer's adjoint buffer. This is semantically correct
and fixes previous miscalculations.

With this change, `AdjointEmitter` visitors for projections are no
longer necessary.

Change usefulness propagation through buffers to be recursive, instead
of stopping after one level.

Resolves TF-246.

eaplatanios committed:
Fixed the '_typeList' derivation and added a derivation for '_unknownShapeList'.

dan-zheng committed:
[AutoDiff] TF-249: Enable `@differentiable` on initializers. (#22846)

`@differentiable` may now be declared on initializers.
Differentiation of initializers is already supported by the
existing differentiation infrastructure.

Resolves TF-249.

Introduces TF-284: unexpected error for unmet `@differentiable`
protocol initializer requirements.

Add comments referencing TF-285: handling multiple missing
`@differentiable` attributes on protocol requirements.

dan-zheng committed:
[Sema] Tweak `Differentiable` derived conformances. (#22850)

By default, `createImplicitConstructor` set the parent context of the implicit
initializer to be the nominal type on which it is declared. However, the
parent context should actually be the conformance context.

This fixes a crash that occurs when the default implementation of
`Differentiable.moved` is removed, so that `moved` is always synthesized.

eaplatanios committed:
Added a derivation for '_unpackTensorHandles'.

eaplatanios committed:
Minor formatting edits.

eaplatanios committed:
Tabs to spaces.

dan-zheng committed:
[AutoDiff] TF-274: Fix usefulness propagation for address-type outputs. (#22852)

* [AutoDiff] TF-274: Fix usefulness propagation for address-type outputs.

Recursively propagate usefulness for address-type outputs, instead of only
marking them as useful.

Resolves TF-274. Update tests.

* Fix/simplify usefulness marking logic.

bgogul committed:
Update tensorflow and tensorflow-swift-bindings commit hashes. (#22789)

A few more changes: 
  * Fix bindings and update test. 
  * Replace Int64 attribute test with randomUniform so that it works on GPU as well.
  * Disable failing tracer_tffunc test in GPU mode.

eaplatanios committed:
Minor changes.

rxwei commented:
Nice! It's awesome that you are pushing this through!

rxwei commented:
> Also, why was there a default implementation for `_unknownShapeList`? I removed it because this prevents the type checker from deriving it.

The default implementation exists because the requirement does not need to be derived, and it actually should not be a customization at all (it's always default-implementable). What should be removed is the protocol requirement, not the default implementation.

rxwei committed:
[AutoDiff] TF-19: Perform activity analysis on all types. (#22848)

* Update Differentiation to do activity analysis on all types.

This allows generating warnings in cases like: `Float(Int(x))` and
other such cases.

Note that this may break some existing semantics by being stricter. In
these cases you can work around it by tagging the function
`@_semantics("autodiff.nonvarying")`.

The user-facing form of this is `Differentiable.withoutDerivative()`.

Resolves [TF-19](https://bugs.swift.org/browse/TF-19) and
[TF-164](https://bugs.swift.org/browse/TF-164).

dan-zheng committed:
[Python] Add `Strideable` conformance, remove slice initializer. (#22849)

Add `Strideable` conformance to `PythonObject`.
This enables `stride(from:to:by:)` and `stride(from:through:by:)`
with `PythonObject`.

Remove unnecessary `init(sliceStart:end:step:)` initializer.
Instead, directly call the builtin `Python.slice()`.

Note: conditionally conforming `StrideTo` and `StrideThrough` to
`PythonConvertible` would enable indexing into `PythonObject` arrays
using them. However, this is difficult because `_stride` is an internal
stored property in `StrideTo` and `StrideThrough`.

dan-zheng committed:
[AutoDiff] TF-277: Fix type-remapping crash. (#22851)

* [AutoDiff] TF-277: Fix type-remapping crash.

Previously, `getOpASTType` was used in `visitLoadInst` and `visitStoreInst` to
remap SIL types that were converted to canonical types. However, this triggered
an assertion failure regarding remapping lowered types.

Now, `PrimalGenCloner::visitLoadInst` and `PrimalGenCloner::visitStoreInst` are removed, because they weren't supposed to be here in the first place if there were no bugs with the cloner.

rxwei committed:
[AutoDiff] Remove the method mode from 'Builtin.autodiffApply'. (#22859)

Before indirect passing was supported, the `_method` configuration in `Builtin.autodiffApply_vjp_method` was used to directly apply a method's VJP as if its arguments were flattened like SIL. Not only was it a workaround, as a trade-off, it also introduced some semantic incorrectness in `@differentiable` functions: It made `@differentiable (T) -> (U) -> V` mean that `U` was also a differentiation parameter in the function, but `@differentiable` should not apply to inner curry levels.

After indirect passing support landed, we removed the method-specific differential operator from the standard library (#22637). So the `_method` mode in `Builtin.autodiffApply` becomes dead code. This PR removes that and simplifies the infrastructure.

eaplatanios committed:
Fixed the derivation for '_unpackTensorHandles'.

eaplatanios commented:
@rxwei Thanks! So, would it be ok if I replace the declaration in the `TensorGroup` protocol with the following?
```swift
static var _unknownShapeList: [TensorShape?] {
  return Array(repeating: nil, count: _typeList.count)
}
```
Also, what is that method used for? Do you always assume unknown shape information when dealing with `TensorGroup`s?

eaplatanios commented:
An even better question is, why not have `_shapeList: [TensorShape]`, where you get the static shape information of the tensors in the group (i.e., through the shape attribute -- if that's supported in S4TF).

eaplatanios committed:
Added a derivation for the 'TensorGroup' initializer.

rxwei commented:
> So, would it be ok if I replace the declaration in the `TensorGroup` protocol with the following?

No, functions in Swift protocols cannot have a body. Default implementations are implemented using protocol extensions, exactly like how `_unknownShapeList` is implemented today. You should just remove the protocol requirement `_unknownShapeList` instead.

> An even better question is, why not have `_shapeList: [TensorShape]`, where you get the static shape information of the tensors in the group (i.e., through the shape attribute -- if that's supported in S4TF).

That isn't what `_unknownShapeList` is used for. It's _literally_ used as an array of `nil`s to represent known a list of unknown shapes to be passed into dataset ops. It would be less ideal and inefficient to use the `shape` property on tensors, because that would bypass the TF runtime unnecessarily while shapes can be inferred through those ops. All we need is to pass a list of `nil`s as a placeholder that also indicates the cardinality of the tensor group.

eaplatanios committed:
Fixed the derivation for the 'TensorGroup' initializer.

eaplatanios committed:
Cleaned up all places where 'TensorGroup's are being used.

eaplatanios committed:
Removed the derivation for '_unknownShapeList'.

eaplatanios commented:
@dan-zheng @rxwei I think the implementation is complete now. I also made the following couple changes:
- Removed the `_unknownShapeList` protocol requirement as @rxwei suggested and added back the default implementation in a `TensorGroup` extension.
- Removed `TensorPair` entirely and just added the following just above the dataset zip operation:
```swift
/// A 2-tuple-like struct that conforms to TensorGroup that represents the 
/// result type of a zip operation between two datasets.
public struct ZippedTensors<T: TensorGroup, U: TensorGroup> : TensorGroup {
  public var first: T
  public var second: U
}
```
That's the only place where it was used and given the automatic derivation I don't think it's an important enough component to have it's own file. I don't know your conventions though and could revert back if you want.

The code is not super nice and I was wondering what your formatting conventions are. I noticed that the C++ codebase uses various styles so I wasn't sure which you prefer. Also, I haven't added tests yet, but I plan to do so soon. However, I'm not sure how to test the initializer and the unpacking function. Is there an example of other tests I could use?

eaplatanios committed:
Minor edit.

rxwei commented:
Great. The additional changes you described above make sense to me. Can you confirm that this is passing all existing `TensorFlow` and `TensorFlowRuntime` tests on your machine locally?

eaplatanios commented:
I'm currently compiling. I'll let you know when I'm able to run them. I was using incremental builds using `ninja swift` so far, and that was great, but when I tried running the tests, the build script decided to recompile pretty much everything. Is there a way to also run tests through ninja?

eaplatanios commented:
I am getting a weird error. So for this piece code in the stdlib, conformance is not derived while compiling:
```
/// A 2-tuple-like struct that conforms to TensorGroup that represents the 
/// result type of a zip operation between two datasets.
public struct ZippedTensors<T : TensorGroup, U : TensorGroup> : TensorGroup {
  public var first: T
  public var second: U
}

// TODO(SR-9156): This does not work in graph mode.
@inlinable @inline(__always)
public func zip<T : TensorGroup, U : TensorGroup>(
  _ dataset1: Dataset<T>, _ dataset2: Dataset<U>
) -> Dataset<ZippedTensors<T, U>> {
  let handle: VariantHandle = #tfop(
     "ZipDataset", ZippedTensors(dataset1._handle, dataset2._handle),
     output_types$dtype: ZippedTensors<T, U>._typeList,
     output_shapes: ZippedTensors<T, U>._unknownShapeList)
  return Dataset(_handle: handle)
}
```
However, when I try to compile the following separately, using the freshly compiled swiftc, everything works fine:
```
public struct TensorPair2<T : TensorGroup, U : TensorGroup> : TensorGroup {
  public var first: T
  public var second: U
}

let test = TensorPair2(
  first: TensorPair2(
    first: Tensor<Int32>([[0, 3], [4, 5]]),
    second: Tensor<Bool>([true, false])), 
  second: Tensor<Float32>([0.0, 0.5]))

print(type(of: test)._typeList)
```
Do you know why this may be happening? Which Swift compiler does the build script use to compile the stdlib?

dan-zheng committed:
[AutoDiff] TF-282: Handle `begin_access` in `getAdjointBuffer`. (#22847)

`begin_access` is not considered a projection, but should be handled
like a projection in `getAdjointBuffer`.

The adjoint buffer for `begin_access` should be a corresponding
`begin_access` into the underlying buffer's adjoint buffer.

Resolves TF-282.

eaplatanios commented:
I can actually verify that if I move `ZippedTensors` and the `zip` function outside of stdlib and compile separately, then everything works fine. However, if I leave them in the stdlib, then I get a compiler error saying that `ZippedTensors` does not conform to `TensorGroup`. Why does that happen?

eaplatanios commented:
I think this may be due to compiling the struct for which the compiler needs to derive conformance, together with the actual `TensorGroup` protocol. I'm not sure what to do to resolve this.

rxwei commented:
Could you share the exact error you got when compiling stdlib?

eaplatanios commented:
If I include the following:
```swift
/// A 2-tuple-like struct that conforms to TensorGroup that represents the 
/// result type of a zip operation between two datasets.
public struct ZippedTensors<T : TensorGroup, U : TensorGroup> : TensorGroup {
  public var first: T
  public var second: U

  public init(_ first: T, _ second: U) {
    self.first = first
    self.second = second
  }
}

// TODO(SR-9156): This does not work in graph mode.
@inlinable @inline(__always)
public func zip<T : TensorGroup, U : TensorGroup>(
  _ dataset1: Dataset<T>, _ dataset2: Dataset<U>
) -> Dataset<ZippedTensors<T, U>> {
  let handle: VariantHandle = #tfop(
     "ZipDataset", ZippedTensors(dataset1._handle, dataset2._handle),
     output_types$dtype: ZippedTensors<T, U>._typeList,
     output_shapes: ZippedTensors<T, U>._unknownShapeList)
  return Dataset(_handle: handle)
}
```
I get this error:
```
swift-source/swift/stdlib/public/TensorFlow/Dataset.swift:192:15: error: type 'ZippedTensors<T, U>' does not conform to protocol 'TensorGroup'
public struct ZippedTensors<T : TensorGroup, U : TensorGroup> : TensorGroup {
              ^
swift-source/swift/stdlib/public/TensorFlow/Dataset.swift:192:15: error: type 'ZippedTensors<T, U>' does not conform to protocol 'TensorArrayProtocol'
public struct ZippedTensors<T : TensorGroup, U : TensorGroup> : TensorGroup {
              ^
swift-source/swift/stdlib/public/TensorFlow/TensorGroup.swift:41:14: note: protocol requires property '_typeList' with type '[TensorDataType]'; do you want to add a stub?
  static var _typeList: [TensorDataType] { get }
             ^
swift-source/swift/stdlib/public/TensorFlow/TensorGroup.swift:45:3: note: protocol requires initializer 'init(_owning:)' with type '(_owning: UnsafePointer<CTensorHandle>?)' (aka '(_owning: Optional<UnsafePointer<OpaquePointer>>)'); do you want to add a stub?
  init(_owning tensorHandles: UnsafePointer<CTensorHandle>?)
  ^
swift-source/swift/stdlib/public/TensorFlow/TensorGroup.swift:29:8: note: protocol requires function '_unpackTensorHandles(into:)' with type '(UnsafeMutablePointer<CTensorHandle>?) -> ()' (aka '(Optional<UnsafeMutablePointer<OpaquePointer>>) -> ()'); do you want to add a stub?
  func _unpackTensorHandles(into address: UnsafeMutablePointer<CTensorHandle>?)
       ^
swift-source/swift/stdlib/public/TensorFlow/Dataset.swift:209:26: error: type 'ZippedTensors<T, U>' has no member '_typeList'
     output_types$dtype: ZippedTensors<T, U>._typeList,
                         ^~~~~~~~~~~~~~~~~~~
swift-source/swift/stdlib/public/TensorFlow/TensorGroup.swift:41:14: note: did you mean '_typeList'?
  static var _typeList: [TensorDataType] { get }
             ^
```
But if I move that code outside of stdlib and compile it independently, it compiles fine.

eaplatanios commented:
I tried to trace this down and I believe that `TensorGroup` is not considered a known protocol by the compiler when the compilation of `ZippedTensors` is triggered. Could this be because it hasn't been compiled yet? And is there a way to avoid this by maybe having the compiler first compile `TensorGroup` and treat it as a known protocol?

eaplatanios commented:
I can now verify this because I also tried adding the following hacky tweak at Line 53 of DerivedConformances.cpp:
```c++
  // Only known protocols can be derived.
  auto knownProtocol = Protocol->getKnownProtocolKind();
  if (Protocol->getNameStr().contains("TensorGroup")) {
    knownProtocol = KnownProtocolKind::TensorGroup;
    Protocol->setKnownProtocolKind(KnownProtocolKind::TensorGroup);
  }
  if (Protocol->getNameStr().contains("TensorArrayProtocol")) {
    knownProtocol = KnownProtocolKind::TensorArrayProtocol;
    Protocol->setKnownProtocolKind(KnownProtocolKind::TensorArrayProtocol);
  }
  if (!knownProtocol)
    return false;
```
When I do that, stdlib compiles fine.

dan-zheng commented:
I wonder what happens if you move `TensorGroup.swift` above `Dataset.swift` in `stdlib/public/TensorFlow/CMakeLists.txt`?

eaplatanios commented:
I'll try that, but looking at the actual invocation to swiftc, all the swift source file names are dumped in a text file in alphabetical order and then that's passed to swiftc. Also, I noticed that the Swift compiler only sets the known protocol flags after loading the standard library (according to my understanding). Thus, there may be no known protocols while combining the standard library.

eaplatanios commented:
@dan-zheng I just tried that and can confirm it doesn't work. :/

eaplatanios commented:
I digged a bit and discovered the file `KnownStdlibTypes.def`. I added `TensorArrayProtocol` and `TensorGroup` there as protocol declarations and I'm testing this now.

eaplatanios commented:
Never-mind that didn't work.

dan-zheng commented:
I found the fix for the "no known protocol found" error:
```diff
--- a/lib/AST/ASTContext.cpp
+++ b/lib/AST/ASTContext.cpp
@@ -1002,6 +1002,7 @@ ProtocolDecl *ASTContext::getProtocol(KnownProtocolKind kind) const {
   for (auto result : results) {
     if (auto protocol = dyn_cast<ProtocolDecl>(result)) {
       getImpl().KnownProtocols[index] = protocol;
+      protocol->setKnownProtocolKind(kind);
       return protocol;
     }
   }
```

`ASTContext::getProtocol` looks up known protocols, but doesn't set the known protocol kind bit in the protocol. This led to the unexpected behavior (known protocol initially found, but not found later by derived conformances).

With this change, the stdlib compiles on my [WIP branch](https://github.com/dan-zheng/swift/tree/anthony). Feel free to cherry-pick the changes - all that's left is to add tests. 🙂

eaplatanios commented:
That's a great find! Thanks a lot Dan! I'll verify that the existing tensorflow tests pass fine and try to add a couple for the derivation later today.

eaplatanios committed:
Set known protocol kind, edit `ZippedTensors`.
- Set known protocol kind in `ASTContext::getProtocol`.
  - Otherwise, unexpected behavior may occur (e.g. a known protocol is found
    in the appropriate module, but later the protocol cannot be found by
    derived conformances).
- Add label-less initializer to `ZippedTensors`.

eaplatanios committed:
Merged upstream changes.

eaplatanios committed:
Resolved merge conflict.

eaplatanios commented:
I incorporated Dan's changes and also made the dataset handle and initializer public so that users can add support for new datasets using raw TF ops from outside stdlib.

eaplatanios committed:
Added some tests for the TensorGroup conformance derivation.

eaplatanios committed:
Fixed the TensorGroup tests.

eaplatanios commented:
I added some tests for TensorGroup and can verify that they pass. Also, all existing tests pass, except for `DatasetAsGlobalVar` and `IteratorAsGlobalVar`, which fail with the following error:
```
Fatal error: GetNext() failed because the iterator has not been initialized. Ensure that you have run the initializer operation for this iterator before getting the next element.
```
This does not seem to be relevant to the TensorGroup conformance derivation though.

eaplatanios committed:
Merge remote-tracking branch 'upstream/tensorflow' into tensorflow

eaplatanios committed:
Bug fixes in the TensorArrayProtocol conformance derivation.

eaplatanios committed:
Merged upstream changes.

eaplatanios commented:
@rxwei @dan-zheng I merged upstream changes and now all existing tests for this PR pass on my local machine.

eaplatanios commented:
I just got back from traveling and was wondering what the status of this issue is. Is there something remaining that I need to do?

dan-zheng commented on lib/Sema/DerivedConformanceTensorGroup.cpp:
```diff
@@ -0,0 +1,396 @@
+  auto *tensorGroupProto = C.getProtocol(KnownProtocolKind::TensorGroup);
+  auto *typeListReq = getProtocolRequirement(tensorGroupProto, C.Id_typeList);
+
+  // Concatenate all member `_typeList` arrays..
```
```suggestion
  // Concatenate all member `_typeList` arrays.
```

dan-zheng commented on lib/Sema/DerivedConformanceTensorGroup.cpp:
```diff
@@ -0,0 +1,396 @@
+
+// Synthesize body for `init(_owning:)`.
+static void 
+deriveBodyTensorGroup_init(
```
Combine into one line:
```suggestion
deriveBodyTensorGroup_init(AbstractFunctionDecl *funcDecl) {
```

dan-zheng commented on test/TensorFlowRuntime/tensor_group.swift:
```diff
@@ -0,0 +1,72 @@
+    return Int(lhs._cDataType.rawValue) == Int(rhs._cDataType.rawValue)
+  }
+}
+
```
Could you please add an empty struct test?
`struct Empty : TensorGroup {}`

eaplatanios replied:
Would that consist of simply adding this empty struct and making sure it compiles, or also checking for whether the data types and shapes are empty arrays?

eaplatanios replied:
I added a simple test just in case and addressed all the feedback. :)

dan-zheng reviewed (approved):
LGTM!

eaplatanios committed:
Addressed Dan's feedback.

eaplatanios committed:
Merge remote-tracking branch 'upstream/tensorflow' into tensorflow

dan-zheng commented on stdlib/public/TensorFlow/TensorGroup.swift:
```diff
@@ -40,21 +40,10 @@ public protocol TensorArrayProtocol {
-  // custom const-evaluable implementations. When the const-evaluator is
-  // powerful enough to evaluate the default implementation, remove this
-  // requirement.
-  static var _unknownShapeList: [TensorShape?] { get }
```
Could you explain why this requirement was removed?

eaplatanios replied:
I removed it based on [this comment](https://github.com/apple/swift/pull/22854#issuecomment-466731011) by @rxwei.

dan-zheng replied:
Ah, thanks!

dan-zheng commented on test/TensorFlowRuntime/tensor_group.swift:
```diff
@@ -0,0 +1,78 @@
+  expectEqual([], Empty._typeList)
+}
+
+TensorGroupTests.test("Simple") {
```
Could you please add some tests for the other derived implementations too?
Namely `init(_owning:)` and `func _unpackTensorHandles(into address: UnsafeMutablePointer<CTensorHandle>?)`.

If derived implementations are correct, we can be confident they'll work with compiler infra for `TensorGroup` types.

eaplatanios replied:
I wanted to test these implementations too, but I'm not sure how. How do I test for memory ownership of tensors? Do you guys have some standard way for doing that?

eaplatanios replied:
@dan-zheng @rxwei I have addressed everything else except this because I'm not sure how I can test for memory ownership.

dan-zheng replied:
I don't think you need to test ownership in particular - if you could please add tests for synthesized implementations of `init(_owning:)` and `func _unpackTensorHandles(into address: UnsafeMutablePointer<CTensorHandle>?)`, that would be great!

Here's an example for `init(_owning):` that passes with current hardcoded `TensorPair`:
```swift
import TensorFlow
let x = Tensor<Float>(1)
let y = Tensor<Float>(2)

let buffer = UnsafeMutableBufferPointer<CTensorHandle>.allocate(capacity: 2)
buffer.baseAddress?.initialize(to: x.handle._cTensorHandle)
buffer.baseAddress?.advanced(by: 1).initialize(to: y.handle._cTensorHandle)

// Initializer signature:
// init(_owning tensorHandles: UnsafePointer<CTensorHandle>?)
let pair = TensorPair<Tensor<Float>, Tensor<Float>>(
  _owning: UnsafePointer(buffer.baseAddress))
expectEqual(TensorPair(Tensor<Float>(1), Tensor<Float>(2)), pair)
```

If you haven't used them before, here are some resources for Swift pointer APIs:
* [`OpaquePointer`](https://developer.apple.com/documentation/swift/opaquepointer): this is the type of `Tensor(1).handle._cTensorHandle`. It's an opaque C pointer.
* [`UnsafePointer<T>`](https://developer.apple.com/documentation/swift/unsafepointer): this is a typed pointer.

eaplatanios replied:
I added the tests for `_unpackTensorHandles`, but having issues with `init(_owning:)`. Actually, even the example you provide for the hardcoded `TensorPair` gives me a segfault. Are you sure that this example runs fine?

eaplatanios replied:
I believe there may be a bug in the base cases for `TensorGroup` in the stdlib. I'm checking now and if indeed that's the case I'll add a fix in this PR.

eaplatanios replied:
Even the following simple example fails with a segfault:

```swift
let x = Tensor<Float>(1)
let buffer = UnsafeMutableBufferPointer<CTensorHandle>.allocate(capacity: 1)
buffer.baseAddress?.initialize(to: x.handle._cTensorHandle)
let t = Tensor<Float>(_owning: UnsafePointer(buffer.baseAddress))
```

dan-zheng replied:
> Actually, even the example you provide for the hardcoded `TensorPair` gives me a segfault. Are you sure that this example runs fine?

I ran my example using the [v0.2 macOS toolchain](https://github.com/tensorflow/swift/blob/master/Installation.md#pre-built-packages), sans `TensorGroup` derived conformances. It seems to work fine for me.

> Even the following simple example fails with a segfault:
That example runs fine for me too.

I wonder what build are you using to test these programs? Is it your development build for this PR?

eaplatanios replied:
It is my development build with the changes from the master branch merged. I don't see how this should affect the single tensor example shown above though that does not use TensorGroups. Is there anything I should be aware of when putting the example code within a test as follows?

```swift
ExampleTests.test("SimpleInit") {
  let x = Tensor<Float>(1)
  let buffer = UnsafeMutableBufferPointer<CTensorHandle>.allocate(capacity: 1)
  buffer.baseAddress?.initialize(to: x.handle._cTensorHandle)
  let t = Tensor<Float>(_owning: UnsafePointer(buffer.baseAddress))
}
```

dan-zheng replied:
Hmm. When I run the program as a standalone Swift file, it works fine:
```console
$ cat tensor.swift
import TensorFlow
let x = Tensor<Float>(1)
let buffer = UnsafeMutableBufferPointer<CTensorHandle>.allocate(capacity: 1)
buffer.baseAddress?.initialize(to: x.handle._cTensorHandle)
let t = Tensor<Float>(_owning: UnsafePointer(buffer.baseAddress))
print(t)
$ $SWIFT_BIN/swift tensor.swift
1.0
```

When I add it as a `TensorFlowRuntime` test and run with `lit` however, it does crash.
```
[ RUN      ] Tensor.SimpleInit
stderr>>> CRASHED: SIGSEGV
the test crashed unexpectedly
[     FAIL ] Tensor.SimpleInit
Tensor: Some tests failed, aborting
UXPASS: []
FAIL: ["SimpleInit"]
SKIP: []
```

Here's what lldb shows:
```
Process 93496 launched: '/Users/danielzheng/swift-dev/build/Ninja-ReleaseAssert+stdlib-Release/swift-macosx-x86_64/test-macosx-x86_64/TensorFlowRuntime/Output/simple.swift.tmp/a.out' (x86_64)
StdlibUnittest: using filter: SimpleInit
[ RUN      ] Tensor.SimpleInit
a.out(93496,0x1000c65c0) malloc: *** error for object 0x10e931050: pointer being freed was not allocated
a.out(93496,0x1000c65c0) malloc: *** set a breakpoint in malloc_error_break to debug
Process 93496 stopped
* thread #1, queue = 'com.apple.main-thread', stop reason = signal SIGABRT
    frame #0: 0x00007fff6be3623e libsystem_kernel.dylib`__pthread_kill + 10
libsystem_kernel.dylib`__pthread_kill:
->  0x7fff6be3623e <+10>: jae    0x7fff6be36248            ; <+20>
    0x7fff6be36240 <+12>: movq   %rax, %rdi
    0x7fff6be36243 <+15>: jmp    0x7fff6be303b7            ; cerror_nocancel
    0x7fff6be36248 <+20>: retq
Target 0: (a.out) stopped.
```

I don't quite have time to investigate this further sadly. My `RUN` line is `%target-run-eager-swift %swift-tensorflow-test-run-extra-options`, I wonder if this includes compiler flags that might affect compilation and cause `pointer being freed was not allocated`.

eaplatanios replied:
Thanks for verifying this so I know it's not something in my setup. I believe there may be a more important underlying issue. In the above example, tensors `x` and `t` are both pointing to the same underlying memory managed by TF. In this case, when we deallocate `x`, for example, is that memory freed? If yes, then when deallocating `t` after that, we may get a segfault. Given the `_owning` argument label I would assume that `t` should own the memory, but then what happens with `x`? Because there is no way for `x` to know that `t` owns the underlying memory. Does this make sense?

dan-zheng replied:
I see your point about the potential "underlying issue".
I wonder if someone more familiar with tensor packing/unpacking can comment on this: cc @marcrasi (who implemented it in https://github.com/apple/swift/pull/20003).
* What are your thoughts on the ownership issue raised by Anthony?
* Why is it that the test program above works when run as a standalone Swift program, but not as a TensorFlowRuntime lit test?

marcrasi replied:
This is wrong:
```
ExampleTests.test("SimpleInit") {
  let x = Tensor<Float>(1)
  let buffer = UnsafeMutableBufferPointer<CTensorHandle>.allocate(capacity: 1)
  buffer.baseAddress?.initialize(to: x.handle._cTensorHandle)
  let t = Tensor<Float>(_owning: UnsafePointer(buffer.baseAddress))
}
```
for exactly the reason @eaplatanios says: `x` and `t` both think that they own the underlying TF memory, so they both try to free it. I think the reason that it doesn't crash as standalone top-level code is that Swift might never run destructors for top-level variables.

To correctly call `init(_owning:)`, you'll need to manually construct some CTensorHandles that aren't owned by anything. Here's a way to do that (I haven't compiled or run this):
```
import CTensorFlow
ExampleTests.test("SimpleInit") {
  let x = Tensor<Float>(1)
  
  let status = TF_NewStatus()
  let cTensorHandle = TFE_TensorHandleCopySharingTensor(x.handle._cTensorHandle, status)
  TF_DeleteStatus(status)

  let buffer = UnsafeMutableBufferPointer<CTensorHandle>.allocate(capacity: 1)
  buffer.baseAddress?.initialize(to: cTensorHandle)
  let t = Tensor<Float>(_owning: UnsafePointer(buffer.baseAddress))
}
```

Hopefully all the existing places in S4TF that make calls to `init(_owning:)` do this correctly :). [Here's an example of a place](https://github.com/apple/swift/blob/0ad31669329e5825b14742b9de11648bc8a1e9eb/stdlib/public/TensorFlow/CompilerRuntime.swift#L791) that appears to do it correctly, so that's good.

eaplatanios replied:
Thanks a lot @marcrasi ! That makes sense. :)

eaplatanios replied:
Added the tests and they now all succeed! :)

rxwei commented on test/TensorFlowRuntime/tensor_group.swift:
```diff
@@ -0,0 +1,78 @@
+  }
+}
+
+struct Empty : TensorGroup {}
```
It would be great if `TensorArrayProtocol` conformances are thoroughly tested too.

dan-zheng replied:
Yes, `TensorArrayProtocol` tests please!
I think this PR will be ready for merging after tests for all derived implementations are added.

eaplatanios replied:
Haha I actually implemented some and discovered I'm missing out the derivation for `_tensorHandleCount` because it was implemented for `TensorGroup` by default. I'll add that later today. :)

rxwei commented on test/TensorFlowRuntime/tensor_group.swift:
```diff
@@ -0,0 +1,78 @@
+
+struct Generic<T: TensorGroup, U: TensorGroup> : TensorGroup {
+  var t: T
+  var u: U
```
Could you add some tests for nested types, for example:

```swift
struct Foo<T: TenosrGroup> {
  func foo<U: TensorGroup>() {
    struct UltraNested<V: TensorGroup> {
      var a: Generic<T, U>
      var b: Generic<U, V>
      var c: Generic<T, V>
    }
  }
}
```

Same for `TensorArrayProtocol`.

cc @dan-zheng. I remember seeing a crash when a nested type uses a generic parameter from the parent function, when writing the custom differentiation tutorial. Would be great to test this for `Differentiable` derived conformances as well.

dan-zheng replied:
> cc @dan-zheng. I remember seeing a crash when a nested type uses a generic parameter from the parent function, when writing the custom differentiation tutorial. Would be great to test this for Differentiable derived conformances as well.

Yes, I remember this bug. It's a latent bug that first existed in `Parameterized` derived conformances, then got copied elsewhere.

It shouldn't be hard to fix - it's a matter of calling `mapTypeIntoContext` on the correct `DeclContext`. I can take a look if the bug surfaces.

eaplatanios replied:
Actually I remember being confused about which `DeclContext` to use in each case, but I thought it through a bit and hope it should not exist now. I'll add some more tests for that.

eaplatanios replied:
Added a couple simple nested generic tests and they seem to work fine.

rxwei replied:
The test case you added is a bit different from the example I gave. In my example `UltraNested` is defined in the body of a generic function. Types nested in function scopes were places where I've seen crashers.

eaplatanios replied:
Yes that was intentional because the example above does not compile. The Swift compiler says that I am not allowed to define a nested struct in a generic function.

rxwei replied:
Can you make the function be non-generic?

eaplatanios replied:
So, even when I do that, how can I refer to that nested struct type from outside the function so that I can test its static properties?

rxwei replied:
Test by calling the function.
```swift
struct Foo<T: TenosrGroup> {
  func foo() {
    struct UltraNested<U: TensorGroup, V: TensorGroup> {
      var a: Generic<T, U>
      var b: Generic<U, V>
      var c: Generic<T, V>
    }
    let x: UltraNested = ...
    expectEqual(...)
  }
}
Foo(...).foo()
```

eaplatanios replied:
Thanks! I just added some tests that are almost what you want. I had to move `T` inside the function because otherwise I get `error: type 'UltraNested' cannot be nested in generic function 'function()'`. If that's good, is there anything else remaining to do?

rxwei replied:
Looks great!

eaplatanios committed:
Fixed a bug in the TensorGroup conformance derivation for properly handling empty structs.

eaplatanios committed:
Added derivation for 'TensorArrayProtocol._tensorHandleCount'.

eaplatanios committed:
Added some tests for the TensorArrayProtocol conformance derivation.

eaplatanios committed:
Added nested type tests for the 'TensorGroup' and 'TensorArrayProtocol' conformance derivations.

eaplatanios committed:
Added tests for 'TensorArrayProtocol._unpackTensorHandles'.

eaplatanios committed:
Added tests for 'TensorGroup.init(_owning:)'.

eaplatanios committed:
Modified the nested generic tests for 'TensorGroup' and 'TensorArrayProtocol'.

rxwei commented on test/TensorFlowRuntime/tensor_group.swift:
```diff
@@ -0,0 +1,230 @@
+  TF_DeleteStatus(status)
+
+  let buffer = UnsafeMutableBufferPointer<CTensorHandle>.allocate(capacity: 4)
+  buffer.baseAddress?.initialize(to: wHandle)
```
`UnsafeMutableBufferPoninter` has its own [`initialize(from:)`](https://developer.apple.com/documentation/swift/unsafemutablebufferpointer), which lets you initialize all four elements from a sequence. 

```swift
buffer.initialize(from: [wHandle, bHanble, floatHandle, intHandle])
```

Same for other occurrences.

rxwei commented on test/TensorFlowRuntime/tensor_group.swift:
```diff
@@ -0,0 +1,230 @@
+TensorGroupTests.test("GenericTypeList") {
+  let float = Float.tensorFlowDataType
+  let int = Int32.tensorFlowDataType
+  expectEqual(
```
Suggested formatting:
```swift
  expectEqual([float, float, float, int],  Generic<Simple, Mixed>._typeList)
```

eaplatanios commented:
I believe this should be ready to be merged. All tests pass on my machine. Could we also run them on the CI server?

rxwei commented on test/TensorFlowRuntime/tensor_group.swift:
```diff
@@ -0,0 +1,230 @@
+      }
+      let float = Float.tensorFlowDataType
+      let int = Int32.tensorFlowDataType
+      expectEqual(
```
Suggested formatting:
```swift
      expectEqual([float, float, float, int, float, int, float, float],
                  UltraNested<Simple, Mixed>._typeList)
```

rxwei commented:
The implementation LGTM. I can trigger tests for you when those comments get addressed. Thanks!

eaplatanios committed:
Addressed Richard's formatting comments.

eaplatanios commented:
Thanks Richard! The comments have now been addressed!

rxwei commented on test/TensorFlowRuntime/tensor_array_protocol.swift:
```diff
@@ -0,0 +1,219 @@
+      let buffer = UnsafeMutablePointer<CTensorHandle>.allocate(capacity: 10)
+      nested._unpackTensorHandles(into: buffer)
+      let expectedBuffer = UnsafeMutableBufferPointer<CTensorHandle>.initialize(
+    from: [w.handle._cTensorHandle, b.handle._cTensorHandle, 
```
Indent by 2 from `let`.

eaplatanios replied:
Sorry about that. Fixed.

eaplatanios committed:
Formatting fix.

rxwei commented on test/TensorFlowRuntime/tensor_array_protocol.swift:
```diff
@@ -195,11 +195,11 @@ TensorArrayProtocolTests.test("NestedGenericUnpackTensorHandles") {
-           float.handle._cTensorHandle, int.handle._cTensorHandle, 
-           w.handle._cTensorHandle, b.handle._cTensorHandle])
+        from: [w.handle._cTensorHandle, b.handle._cTensorHandle, 
+              string.handle._cTensorHandle, float.handle._cTensorHandle, 
```
Sorry for another nit: This should align with the first element, not with `[`.

eaplatanios replied:
Np. Thanks for being thorough.

eaplatanios committed:
Formatting fix.

rxwei commented:
@swift-ci please test tensorflow

rxwei commented:
@swift-ci please test tensorflow

dan-zheng commented:
@eaplatanios: could you please address the CI test failures? They look like:
```
/home/swiftci/jenkins/workspace/swift-PR-TensorFlow-Linux/swift/test/TensorFlowRuntime/tensor_array_protocol.swift:57:66: error: instance member 'initialize' cannot be used on type 'UnsafeMutableBufferPointer<CTensorHandle>' (aka 'UnsafeMutableBufferPointer<OpaquePointer>'); did you mean to use a value of this type instead?
  let expectedBuffer = UnsafeMutableBufferPointer<CTensorHandle>.initialize(
                       ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ ^
```

eaplatanios committed:
Bug fixes.

eaplatanios commented:
@dan-zheng Sorry I didn't run the tests after the last formatting edits. Fixed now and. checked that the tests pass on my machine.

rxwei commented:
@swift-ci please test tensorflow

rxwei commented:
@swift-ci please test tensorflow

rxwei commented:
@swift-ci please test tensorflow

rxwei commented:
@swift-ci please test tensorflow

rxwei commented:
@swift-ci please test tensorflow

rxwei commented on stdlib/public/TensorFlow/Dataset.swift:
```diff
@@ -185,14 +185,27 @@ extension DatasetIterator : IteratorProtocol {
 }
 
+/// A 2-tuple-like struct that conforms to TensorGroup that represents the 
+/// result type of a zip operation between two datasets.
```
This API is more general than datasets and thus should not mention "dataset" in its documentation.

eaplatanios replied:
I’ll edit this to reflect your comment.

rxwei commented on stdlib/public/TensorFlow/Dataset.swift:
```diff
@@ -185,14 +185,27 @@ extension DatasetIterator : IteratorProtocol {
+/// A 2-tuple-like struct that conforms to TensorGroup that represents the 
+/// result type of a zip operation between two datasets.
+@_fixed_layout
+public struct ZippedTensors<T : TensorGroup, U : TensorGroup> : TensorGroup {
```
I feel the name of this struct should be closer to [`Zip2Sequence`](https://developer.apple.com/documentation/swift/zip2sequence). Perhaps `Zip2Tensors`?

eaplatanios replied:
`Zip2Tensors` sounds good!

rxwei commented:
I'm not sure what's wrong with Linux CPU tests. Perhaps there's a segfault somewhere.

eaplatanios commented:
I have seen these tests failing on my machine too before but wasn’t sure why. I believe I mentioned it in an earlier comment in this thread. I can look into it tomorrow, but I remember I didn’t manage to figure it out before, although I didn’t spend too much time on it.

eaplatanios commented:
@rxwei Not sure which tests fail on the CI server. When I click on the failed test suites, the log shows all of them as having succeeded.

dan-zheng commented:
> @rxwei Not sure which tests fail on the CI server. When I click on the failed test suites, the log shows all of them as having succeeded.

Yes, that's weird. [Two tests crashed](https://ci-external.swift.org/job/swift-PR-TensorFlow-Linux/1933/console) despite printing `All tests passed`.
```
Failing Tests (2):
    Swift(linux-x86_64) :: TensorFlowRuntime/dataset_global.swift
    Swift(linux-x86_64) :: TensorFlowRuntime/dataset_api.swift
```

Retriggering CI to see if it's reproducible.

dan-zheng commented:
@swift-ci Please test tensorflow

eaplatanios commented:
@dan-zheng @rxwei I noticed something interesting about the failing tests. If I try to run the tests using `lit test-macosx-x86_64/TensorFlowRuntime/dataset_api.swift -sv`,  for example, then I see `Unsupported Tests : 1` and no failures. However, when I run all tests together using `./utils/build-script --preset=tensorflow_test`, I get the failures. Is there a way to just run the failing tests without having to rerun everything? Also, this behavior seems unexpected anyway so maybe there is something wrong.

eaplatanios commented:
I tried to dig into this a bit more. It seems as if tests like `DatasetAPI.SingleValueManualIterator_CPU` are run twice. The first time they succeed and the second time they fail with `Fatal error: GetNext() failed because the iterator has not been initialized. Ensure that you have run the initializer operation for this iterator before getting the next element.: file /Users/eaplatanios/Development/GitHub/swift-source/swift/stdlib/public/TensorFlow/CompilerRuntime.swift, line 2035`. Not sure why this happens.

The output looks like this:

```
Command Output (stdout):
--
[ RUN      ] DatasetAPI.SingleValueManualIterator_CPU
[       OK ] DatasetAPI.SingleValueManualIterator_CPU
[ RUN      ] DatasetAPI.DatasetIteration_CPU
[       OK ] DatasetAPI.DatasetIteration_CPU
[ RUN      ] DatasetAPI.SingleValueTransformations_CPU
[       OK ] DatasetAPI.SingleValueTransformations_CPU
[ RUN      ] DatasetAPI.SingleValueHOFs_CPU
[       OK ] DatasetAPI.SingleValueHOFs_CPU
[ RUN      ] DatasetAPI.MapToDifferentType_CPU
[       OK ] DatasetAPI.MapToDifferentType_CPU
[ RUN      ] DatasetAPI.SingleValueBatched_CPU
[       OK ] DatasetAPI.SingleValueBatched_CPU
[ RUN      ] DatasetAPI.DoubleValueDatasetIteration_CPU
[       OK ] DatasetAPI.DoubleValueDatasetIteration_CPU
DatasetAPI: All tests passed
[ RUN      ] DatasetAPI.SingleValueManualIterator_CPU
stderr>>> Fatal error: GetNext() failed because the iterator has not been initialized. Ensure that you have run the initializer operation for this iterator before getting the next element.: file /Users/eaplatanios/Development/GitHub/swift-source/swift/stdlib/public/TensorFlow/CompilerRuntime.swift, line 2035
stderr>>> CRASHED: SIGILL
the test crashed unexpectedly
[     FAIL ] DatasetAPI.SingleValueManualIterator_CPU
[ RUN      ] DatasetAPI.DatasetIteration_CPU
stderr>>> Fatal error: GetNext() failed because the iterator has not been initialized. Ensure that you have run the initializer operation for this iterator before getting the next element.: file /Users/eaplatanios/Development/GitHub/swift-source/swift/stdlib/public/TensorFlow/CompilerRuntime.swift, line 2035
stderr>>> CRASHED: SIGILL
the test crashed unexpectedly
[     FAIL ] DatasetAPI.DatasetIteration_CPU
[ RUN      ] DatasetAPI.SingleValueTransformations_CPU
stderr>>> Fatal error: GetNext() failed because the iterator has not been initialized. Ensure that you have run the initializer operation for this iterator before getting the next element.: file /Users/eaplatanios/Development/GitHub/swift-source/swift/stdlib/public/TensorFlow/CompilerRuntime.swift, line 2035
stderr>>> CRASHED: SIGILL
the test crashed unexpectedly
[     FAIL ] DatasetAPI.SingleValueTransformations_CPU
[ RUN      ] DatasetAPI.SingleValueBatched_CPU
stderr>>> Fatal error: GetNext() failed because the iterator has not been initialized. Ensure that you have run the initializer operation for this iterator before getting the next element.: file /Users/eaplatanios/Development/GitHub/swift-source/swift/stdlib/public/TensorFlow/CompilerRuntime.swift, line 2035
stderr>>> CRASHED: SIGILL
the test crashed unexpectedly
[     FAIL ] DatasetAPI.SingleValueBatched_CPU
DatasetAPI: Some tests failed, aborting
```

rxwei commented:
`TensorFlowRuntime` tests require special flags to `lit`. 

```bash
lit test-macosx-x86_64/TensorFlowRuntime/tensor_autodiff_runtime.swift --param swift_tensorflow_path=../../../../swift-tf-source/tensorflow/bazel-bin/tensorflow --param swift_test_mode=optimize -sv
```

eaplatanios commented:
@rxwei so I found that adding this parameter `--param swift_test_mode=optimize` makes the tests run and fail. Without it they don't even run. Why is that?

eaplatanios commented:
I'm wondering if the `-O` compilation mode is removing the `#tfop("MakeIterator", _handle, resource) as Void` call in `Dataset.makeIterator()`, which is placed there for its side-effects.

rxwei commented:
@marcrasi implemented the eager dataset code path and agreed to take a look.

marcrasi commented:
I'll quickly explain what "--param swift_test_mode=optimize" does, and then I'll take a look at the failure.

It makes the compiler run with `-O`, as you have guessed. [Here's](https://github.com/apple/swift/blob/35f5c7afb3f54caba3a14cc71dbc9bb3a04ecebc/test/lit.cfg#L515) where it does that.

It also enables tests that have a line saying `// REQUIRES: swift_test_mode_optimize`. So that's why the test doesn't run when you don't specify it. (A lot of the TF tests uses to depend on `-O`. I think recent developments may have made them run without `-O`, but looks like we haven't gotten around to removing `// REQUIRES: swift_test_mode_optimize` from a lot of places yet).

The CI tests run all the tests once without optimize, and then once again with optimize.

eaplatanios commented:
Thanks for the explanation @marcrasi ! I wasn't aware of the `// REQUIRES: swift_test_mode_optimize` flag.

I noticed that if I remove the `// REQUIRES: swift_test_mode_optimize` from `dataset_api.swift` the tests now run (as expected), but still fail with the same errors, which would suggest that the compiler `-O` flag is not the issue.

marcrasi commented:
> I noticed that if I remove the // REQUIRES: swift_test_mode_optimize from dataset_api.swift the tests now run (as expected), but still fail with the same errors, which would suggest that the compiler -O flag is not the issue.

Good to know, thanks!

Sorry, I haven't had time to actually look at the failure yet. I think I'll have some time in the next few hours.

marcrasi commented:
Note that there are two `RUN` lines in the tests. The first one runs in "eager" mode and the second runs in "graph program extraction" mode. It fails on "graph program extraction" mode.

GPE is related to const-evaluation, and the comment above `static var _unknownShapeList: [TensorShape?] { get }` in `TensorGroup.swift` says something about const-evaluation. So I decided to try adding that line back. It fixed the tests!

If the problem was indeed a problem with const-evaluation, I would have expected the failure message to be something about "expression not constant evaluable". Maybe there is some recent bug in GPE causing it to do weird things when expressions are not const evaluable?

@eaplatanios is this enough information for you to continue?

rxwei commented:
It is likely that there's logic that requires `_unknownShapeList` to be a protocol requirement in addition to having a default implementation in a protocol extension.

eaplatanios commented:
@marcrasi Thanks a lot! That's very informative and useful. So, the question now is: should I just add the protocol requirement back and then we try to merge this PR? Or should I also look into why removing it results in these errors?

rxwei commented:
Feel free to add the requirement back!

marcrasi commented:
Yeah, if the `_unknownShapeList` requirement is compatible with your changes in this PR, adding it back without worrying about why it's necessary sounds good to me.

rxwei commented:
Related note: If this is related to constant evaluation, I suggest you try making `_unknownShapeList` be `@inlinable` first.

eaplatanios committed:
Added back the 'TensorGroup._unknownShapeList' protocol requirement to avoid some test failures.

eaplatanios committed:
Renamed 'ZippedTensors' to 'Zip2Tensor', in order to match the 'Zip2Sequence' naming convention.

eaplatanios commented:
Thanks! I added back the protocol requirement because making the static property in the extension `@inlinable` did not work (it resulted in the same errors). I also renamed `ZippedTensors` to `Zip2Tensor` as per Richard's comment, to match the `Zip2Sequence` convention.

I believe this should be ready to merge. Could we run the tests on the CI again to make sure everything is fine now?

rxwei commented:
Thanks! 

> I also renamed `ZippedTensors` to `Zip2Tensor` as per Richard's comment, to match the `Zip2Sequence` convention.

Nit: My suggestion was actually `Zip2Tensors` because I think it isn't really a `Tensor` but a high dimensional array of pairs. What do you think? Is there a more accurate name for this?

eaplatanios commented:
Oh right. I guess if we want to be consistent we could also go with `Zip2TensorGroup`. It's a bit long but I don't imagine it being frequently spelled out in users' code.

eaplatanios committed:
Renamed 'Zip2Tensor' to 'Zip2TensorGroup'.

eaplatanios commented:
Done! Should hopefully pass all tests now. :)

rxwei commented:
@swift-ci please test tensorflow

rxwei commented:
@swift-ci please test tensorflow

rxwei commented:
@swift-ci please test tensorflow

rxwei commented:
@swift-ci please test tensorflow

rxwei commented:
@swift-ci please test tensorflow

rxwei commented:
@swift-ci please test tensorflow Linux

rxwei commented:
@swift-ci please test tensorflow Linux CPU

rxwei commented:
Thanks Anthony. Congrats on getting this done!

rxwei commented:
Thanks Anthony. Congrats on getting this done!Thanks Anthony. Congrats on getting this done!

eaplatanios commented:
Thanks Richard! And thanks everyone for the helpful pointers and patience. :)
